### ISSUE TITLE
Calling a JS function as a constructor from Java?

### ISSUE THREAD

**Issue by johnsoft** [eclipsesource/J2V8#195]
Let's say I have a reference to the built-in `Date` function:

    V8Function dateConstructor = (V8Function) v8.get("Date");

From the Java side, how could I call this as a constructor `new Date()` instead of calling it as a function `Date()`?

Sorry if this is documented somewhere and I missed it.

**Comment by johnsoft** [comment]
Here is the monstrosity I implemented to work around this. It uses a global helper function called e.g. `__ctor3` that forwards the call. It's Kotlin, not Java, but hopefully it's understandable.

    fun V8Function.callNew(parameters: V8Array?): V8Object {
        val v8 = this.runtime

        val count = if (parameters == null) 0 else parameters.length()
        val funcName = "__ctor$count"
        val paramNames = (0 until count).map { "p" + it }
        val params = (listOf("ctor") + paramNames).joinToString(",")
        val args = paramNames.joinToString(",")
        v8.executeVoidScript("function $funcName($params) { return new ctor($args); }")

        val arguments = if (parameters != null)
            V8Array(v8).push(this).executeArrayFunction("concat", parameters)
        else
            V8Array(v8).push(this)
        return v8.executeObjectFunction("$funcName", arguments)
    }

How to use, given `jsClass` of type `V8Function`:

    jsClass.callNew(arguments)  // translates to JavaScript `new jsClass(...arguments)`

I hope this atrocity prods someone to tell me I missed something easy and there's a better way to do this!

**Comment by nbransby** [comment]
What about:
V8Function dateConstructor = (V8Function) v8.get("Date");
V8Object prototype = dateConstructor .getObject("prototype");
V8Object object = new V8Object(v8);
object.setPrototype(prototype);
object.add("constructor", dateConstructor);
constructor.call(object, new V8Array(v8));

sorry I didn't try myself yet.

**Comment by Haringat** [comment]
@nbransby: That will only work for constructors which do not insist on being "new'ed". While this includes almost everything there are edge cases where it will not work. You could not, for example, create a proxy with this method (because Proxy() will throw an exception telling you that it has to be called with the "new" operator.

For now I helped myself doing with the following:
```Javascript
Object.spawn = function(constructor, ...arguments){
    return new constructor(...arguments);
};
```
Usage (here: created a proxy):
```Java
V8 v8 = V8.createRuntime();
V8Object object = v8.getObject("Object");
V8Object proxy = v8.getObject("Proxy");
V8Object target = new V8Object(this.v8);
V8Object handler = new V8Object(this.v8);
handler.registerJavaMethod(new JavaCallback() {
    @Override
    public Object invoke(V8Object receiver, V8Array parameters) {
        return "It works!";
    }
}, "get");
V8Object result = object.executeObjectFunction("spawn", proxy, target, handler);
String proof = result.getString("foo");
```
While I must admit that this is a bit clunky at usage I think there currently is no elegant way of doing this while the new operator is not implemented in the api.

**Comment by JessHolle** [comment]
I ended up going with:

    public static V8Function getNewInstanceFunction(V8 v8) {
        return (V8Function) v8.executeObjectScript("(function() { return function(clazz) { return new clazz(...Array.prototype.slice.call(arguments, 1)); } })();",
                MyClass.class.getSimpleName() + "_java.js", 0);
    }

I can then hold onto V8Function and re-use it (automating cleanup of it via V8.registerResource).

This is such a common need, however, that I believe J2V8 should provide a higher-level facility like newInstance(V8Object clazz, V8Array args), newInstance(V8Object, Object... args), or the like -- even if the internal implementation ends up being that above.

**Comment by irbull** [comment]
I think that may be an interesting API to try and provide. I'm going to mark this as `good first issue` as I think adding this would allow someone to touch a few different parts of the system, and I don't think providing this API would break any other APIs.

**Comment by JessHolle** [comment]
So what I don't get is whether there is a better way to do this via appropriate, underlying V8 APIs.

If not, moving the utility I note above into the Java layer shouldn't too hard to do in a somewhat nice fashion.  (As it is now I have a map from such V8 instances to instances of this function so I can just them as need without polluting the JS globals.)

**Comment by irbull** [comment]
Yes, I'm not sure either. I have a new person that's going to be helping with this project. I wanted to flag a few issues that I thought would be good to look at. Ones that are non-critical, but could help them get their feet wet in different parts of the system. This seems like a nice one because they can explore the `v8.h` file a bit to see if V8 provides anything useful.
